Init does not work with yarn workspaces

Copy from https://github.com/facebook/react-native/issues/24202 and there's a related issue https://github.com/react-native-community/react-native-cli/issues/271

## 🐛 Bug Report
I'm using `react-native init` to create a simple app inside a monorepo, but the cli seems to have problems with the yarn workspaces (with hoisting) - I'm also using lerna, but i don't think that lerna is causing the problems here.

## To Reproduce
I'm just running:
`react-native init KeepfyApp`

The log is:
```
This will walk you through creating a new React Native project in /home/gabriel/keepfy/root/packages/KeepfyApp
Using yarn v1.15.2
Installing react-native...
yarn add v1.15.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
**========= SOME WARNINGS NOT INCLUDED HERE =========**
[4/4] Building fresh packages...

success Saved 1 new dependency.
info Direct dependencies
info All dependencies
└─ react-native@0.59.2
Done in 12.57s.
internal/modules/cjs/loader.js:651
    throw err;
    ^

Error: Cannot find module '/home/gabriel/keepfy/root/packages/KeepfyApp/node_modules/react-native/package.json'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:649:15)
    at Function.Module._load (internal/modules/cjs/loader.js:575:25)
    at Module.require (internal/modules/cjs/loader.js:705:19)
    at require (internal/modules/cjs/helpers.js:14:16)
    at checkNodeVersion (/home/gabriel/.fnm/node-versions/v11.12.0/installation/lib/node_modules/react-native-cli/index.js:306:21)
    at run (/home/gabriel/.fnm/node-versions/v11.12.0/installation/lib/node_modules/react-native-cli/index.js:300:3)
    at createProject (/home/gabriel/.fnm/node-versions/v11.12.0/installation/lib/node_modules/react-native-cli/index.js:249:3)
    at init (/home/gabriel/.fnm/node-versions/v11.12.0/installation/lib/node_modules/react-native-cli/index.js:200:5)
    at Object.<anonymous> (/home/gabriel/.fnm/node-versions/v11.12.0/installation/lib/node_modules/react-native-cli/index.js:153:7)
    at Module._compile (internal/modules/cjs/loader.js:799:30)
```
The `Error: Cannot find module '/home/gabriel/keepfy/root/packages/KeepfyApp/node_modules/react-native/package.json'` is correct, there's no `node_modules` folder there, they are installed in the root folder, the correct path should be `/home/gabriel/keepfy/root/node_modules/react-native/package.json`


## Expected Behavior

I'm expecting the app to be created, but only the `package.json` is created.

## Code Example

```
// lerna.json
{
  "packages": [
    "packages/*"
  ],
  "npmClient": "yarn",
  "useWorkspaces": true,
  "version": "independent"
}

//package.json (root)
{
  "name": "root",
  "private": true,
  "devDependencies": {
    "lerna": "3.13.1",
    "yarn": "1.15.2"
  },
  "workspaces": [
    "packages/*"
  ],
  "scripts": {
    "build": "lerna run build --stream",
    "bootstrap": "lerna bootstrap"
  }
}
```
With those files in a root project folder, run `cd packages && react-native init` and you should see the error.

There's something to note here, i could start a ''correct'' project only once, `react-native init` created the folder and stuff inside, but it installed ignoring the yarn workspaces (because it created the `yarn.lock` inside the app folder).. so i think that time doesn't count.

## Environment
`[skip envinfo]`

Ran inside the created folder `KeepfyApp`:
`react-native info` ->
```
Command `info` unrecognized. Make sure that you have run `npm install` and that you are inside a react-native project.
```
`react-native --version` ->
```
react-native-cli: 2.0.1
react-native: n/a - not inside a React Native project directory
```

If there's another info i can provide to help just ask.

--- a/packages/cli/src/commands/init/__tests__/template.test.js
+++ b/packages/cli/src/commands/init/__tests__/template.test.js
@@ -12,6 +12,7 @@ import {
 import * as copyFiles from '../../../tools/copyFiles';
 
 const TEMPLATE_NAME = 'templateName';
+const TEMPLATE_SOURCE_DIR = '/tmp/rncli-init-template-123456';
 
 afterEach(() => {
   jest.restoreAllMocks();
@@ -21,17 +22,18 @@ afterEach(() => {
 test('installTemplatePackage', async () => {
   jest.spyOn(PackageManger, 'install').mockImplementationOnce(() => {});
 
-  await installTemplatePackage(TEMPLATE_NAME, true);
+  await installTemplatePackage(TEMPLATE_NAME, TEMPLATE_SOURCE_DIR, true);
 
   expect(PackageManger.install).toHaveBeenCalledWith([TEMPLATE_NAME], {
     preferYarn: false,
     silent: true,
+    cwd: TEMPLATE_SOURCE_DIR,
   });
 });
 
 test('getTemplateConfig', () => {
   jest.mock(
-    `node_modules/${TEMPLATE_NAME}/template.config`,
+    `${TEMPLATE_SOURCE_DIR}/node_modules/${TEMPLATE_NAME}/template.config`,
     () => ({
       placeholderName: 'someName',
       templateDir: 'someDir',
@@ -42,11 +44,12 @@ test('getTemplateConfig', () => {
   );
   jest.spyOn(path, 'resolve').mockImplementationOnce((...e) => e.join('/'));
 
-  expect(getTemplateConfig(TEMPLATE_NAME)).toEqual({
+  expect(getTemplateConfig(TEMPLATE_NAME, TEMPLATE_SOURCE_DIR)).toEqual({
     placeholderName: 'someName',
     templateDir: 'someDir',
   });
   expect(path.resolve).toHaveBeenCalledWith(
+    TEMPLATE_SOURCE_DIR,
     'node_modules',
     TEMPLATE_NAME,
     'template.config',
@@ -61,9 +64,10 @@ test('copyTemplate', () => {
   jest.spyOn(copyFiles, 'default').mockImplementationOnce(() => {});
   jest.spyOn(process, 'cwd').mockImplementationOnce(() => CWD);
 
-  copyTemplate(TEMPLATE_NAME, TEMPLATE_DIR);
+  copyTemplate(TEMPLATE_NAME, TEMPLATE_DIR, TEMPLATE_SOURCE_DIR);
 
   expect(path.resolve).toHaveBeenCalledWith(
+    TEMPLATE_SOURCE_DIR,
     'node_modules',
     TEMPLATE_NAME,
     TEMPLATE_DIR,
@@ -77,9 +81,10 @@ test('executePostInitScript', async () => {
 
   jest.spyOn(path, 'resolve').mockImplementationOnce(() => RESOLVED_PATH);
 
-  await executePostInitScript(TEMPLATE_NAME, SCRIPT_PATH);
+  await executePostInitScript(TEMPLATE_NAME, SCRIPT_PATH, TEMPLATE_SOURCE_DIR);
 
   expect(path.resolve).toHaveBeenCalledWith(
+    TEMPLATE_SOURCE_DIR,
     'node_modules',
     TEMPLATE_NAME,
     SCRIPT_PATH,

--- a/packages/cli/src/commands/init/init.js
+++ b/packages/cli/src/commands/init/init.js
@@ -1,4 +1,6 @@
 // @flow
+import os from 'os';
+import path from 'path';
 import fs from 'fs-extra';
 import minimist from 'minimist';
 import semver from 'semver';
@@ -49,16 +51,18 @@ async function createFromExternalTemplate(
   const loader = new Loader({text: 'Downloading template'});
   loader.start();
 
+  const templateSourceDir = fs.mkdtempSync(
+    path.join(os.tmpdir(), 'rncli-init-template-'),
+  );
   try {
     let {uri, name} = await processTemplateName(templateName);
-
-    await installTemplatePackage(uri, npm);
+    await installTemplatePackage(uri, templateSourceDir, npm);
     loader.succeed();
     loader.start('Copying template');
 
     name = adjustNameIfUrl(name);
-    const templateConfig = getTemplateConfig(name);
-    copyTemplate(name, templateConfig.templateDir);
+    const templateConfig = getTemplateConfig(name, templateSourceDir);
+    copyTemplate(name, templateConfig.templateDir, templateSourceDir);
 
     loader.succeed();
     loader.start('Preparing template');
@@ -70,7 +74,7 @@ async function createFromExternalTemplate(
     if (postInitScript) {
       // Leaving trailing space because there may be stdout from the script
       loader.start('Executing post init script ');
-      await executePostInitScript(name, postInitScript);
+      await executePostInitScript(name, postInitScript, templateSourceDir);
       loader.succeed();
     }
 
@@ -80,6 +84,8 @@ async function createFromExternalTemplate(
   } catch (e) {
     loader.fail();
     throw new Error(e);
+  } finally {
+    fs.removeSync(templateSourceDir);
   }
 }
 
@@ -94,7 +100,9 @@ async function createFromReactNativeTemplate(
   const Loader = getLoader();
   const loader = new Loader({text: 'Downloading template'});
   loader.start();
-
+  const templateSourceDir = fs.mkdtempSync(
+    path.join(os.tmpdir(), 'rncli-init-template-'),
+  );
   try {
     if (semver.valid(version) && !semver.gte(version, '0.60.0')) {
       throw new Error(
@@ -106,13 +114,13 @@ async function createFromReactNativeTemplate(
 
     const {uri} = await processTemplateName(`${TEMPLATE_NAME}@${version}`);
 
-    await installTemplatePackage(uri, npm);
+    await installTemplatePackage(uri, templateSourceDir, npm);
 
     loader.succeed();
     loader.start('Copying template');
 
-    const templateConfig = getTemplateConfig(TEMPLATE_NAME);
-    copyTemplate(TEMPLATE_NAME, templateConfig.templateDir);
+    const templateConfig = getTemplateConfig(TEMPLATE_NAME, templateSourceDir);
+    copyTemplate(TEMPLATE_NAME, templateConfig.templateDir, templateSourceDir);
 
     loader.succeed();
     loader.start('Processing template');
@@ -123,7 +131,11 @@ async function createFromReactNativeTemplate(
     const {postInitScript} = templateConfig;
     if (postInitScript) {
       loader.start('Executing post init script');
-      await executePostInitScript(TEMPLATE_NAME, postInitScript);
+      await executePostInitScript(
+        TEMPLATE_NAME,
+        postInitScript,
+        templateSourceDir,
+      );
       loader.succeed();
     }
 
@@ -133,6 +145,8 @@ async function createFromReactNativeTemplate(
   } catch (e) {
     loader.fail();
     throw new Error(e);
+  } finally {
+    fs.removeSync(templateSourceDir);
   }
 }
 

--- a/packages/cli/src/commands/init/template.js
+++ b/packages/cli/src/commands/init/template.js
@@ -1,5 +1,4 @@
 // @flow
-
 import execa from 'execa';
 import path from 'path';
 import * as PackageManager from '../../tools/packageManager';
@@ -12,16 +11,25 @@ export type TemplateConfig = {
   postInitScript?: string,
 };
 
-export function installTemplatePackage(templateName: string, npm?: boolean) {
+export function installTemplatePackage(
+  templateName: string,
+  cwd: string,
+  npm?: boolean,
+) {
   logger.debug(`Installing template from ${templateName}`);
   return PackageManager.install([templateName], {
     preferYarn: !npm,
     silent: true,
+    cwd,
   });
 }
 
-export function getTemplateConfig(templateName: string): TemplateConfig {
+export function getTemplateConfig(
+  templateName: string,
+  templateSourceDir: string,
+): TemplateConfig {
   const configFilePath = path.resolve(
+    templateSourceDir,
     'node_modules',
     templateName,
     'template.config',
@@ -32,8 +40,17 @@ export function getTemplateConfig(templateName: string): TemplateConfig {
   return require(configFilePath);
 }
 
-export function copyTemplate(templateName: string, templateDir: string) {
-  const templatePath = path.resolve('node_modules', templateName, templateDir);
+export function copyTemplate(
+  templateName: string,
+  templateDir: string,
+  templateSourceDir: string,
+) {
+  const templatePath = path.resolve(
+    templateSourceDir,
+    'node_modules',
+    templateName,
+    templateDir,
+  );
 
   logger.debug(`Copying template from ${templatePath}`);
 
@@ -43,8 +60,14 @@ export function copyTemplate(templateName: string, templateDir: string) {
 export function executePostInitScript(
   templateName: string,
   postInitScript: string,
+  templateSourceDir: string,
 ) {
-  const scriptPath = path.resolve('node_modules', templateName, postInitScript);
+  const scriptPath = path.resolve(
+    templateSourceDir,
+    'node_modules',
+    templateName,
+    postInitScript,
+  );
 
   logger.debug(`Executing post init script located ${scriptPath}`);
 

--- a/packages/cli/src/tools/packageManager.js
+++ b/packages/cli/src/tools/packageManager.js
@@ -6,6 +6,7 @@ import {getYarnVersionIfAvailable, isProjectUsingYarn} from './yarn';
 type Options = {|
   preferYarn?: boolean,
   silent?: boolean,
+  cwd?: string,
 |};
 
 let projectDir;
@@ -18,6 +19,7 @@ function executeCommand(
   return execa(command, args, {
     stdio:
       options && options.silent && !logger.isVerbose() ? 'pipe' : 'inherit',
+    cwd: options && options.cwd,
   });
 }
 

